13 add target data getters

crispy-wonton opened the pull request:
Fixes #13 

# Description

Changes:
- add new `get_target.py` file with getter functions to load and prepare target datasets for 4 features: tenure, property_type, build_year, number_of_rooms
- add new function to `base_getters.py` to enable loading of excel files from s3 with polars
- update base config with target data sources plus additional mapping params
- add new data sources to `config/README.md` 

# Instructions for Reviewer

Please could you:
- check for general bugs or mistakes
- check the docstrings make sense
- check the functions output the expected dataframes for you. For each feature, the dataframes should have one row per LSOA, with one column per feature category. Each cell should contain the corresponding count of that category for the given LSOA

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
add get_target.py with getters for target data for reweighting
getters added for 4 features: tenure, property_type, build_year, number_of_rooms
add new base_getter function to load excel files from s3 with polars

crispy-wonton committed:
add target data sources and mappings to base config file

crispy-wonton committed:
remove unused file from base config and update config data source key for consistency

crispy-wonton committed:
add new data sources to config/README

crispy-wonton committed:
correct config key in config/README

crispy-wonton committed:
add source URLs to config/README

crispy-wonton committed:
add data source descriptions to config/README

crispy-wonton commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,181 @@
+    df = (
+        df.with_columns(
+            [
+                pl.sum_horizontal(pre_cols).alias("pre"),
+                pl.sum_horizontal(post_cols).alias("post"),
+            ]
```
```suggestion
                pl.sum_horizontal(pre_cols).alias("pre_1930"),
                pl.sum_horizontal(post_cols).alias("post_1930"),
            ]
```

crispy-wonton committed:
fix error in get_target.get_df_target_build_year - column name mismatch

crispy-wonton committed:
add pipeline/reweight_epc/prepare_target.py with functions to get target marginals and generate artificial target population with balance

crispy-wonton committed:
round proportions to increase speed of artificial target population generation

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+
+def get_df_target_nrooms() -> pl.DataFrame:
+    """
+    Get dataframe of counts of total number of rooms for properties in all LSOAs in England and Wales. Source: census data 2021.
```
Just add in a brief description of the operations being performed (dropping and renaming)!

crispy-wonton replied:
have added detail about why '9' is renamed to '9+'. For conciseness I'll leave out further descriptions of specific processing as I think it isn't relevant to the output and the code documents itself

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    Returns:
+        pl.Dataframe: counts of total number of rooms for properties in all LSOAs in England and Wales
+    """
+    df = pl.read_csv(config["data_source"]["EW_census_number_of_rooms"])
```
Would it be worth chaining the polar operations together?

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    df = pl.read_csv(config["data_source"]["EW_census_number_of_rooms"])
+    df = df.drop(
+        [
+            "Lower layer Super Output Areas",
```
Probably nice to define these in a a list e.g.: 
`    # Drop unnecessary columns
    columns_to_drop = [
        "Lower layer Super Output Areas",
        "Number of rooms (Valuation Office Agency) (9 categories)"
    ]`

crispy-wonton replied:
I left as is to keep a shorter function, same with dict below

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+            "Number of rooms (Valuation Office Agency) (9 categories) Code": "number_of_rooms",
+        }
+    )
+    return df.pivot(
```
Would suggest not performing an operation in the return statement and have it as a separate operation!

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    constant. Source: census data 2021.
+
+    Args:
+        censored_values (int): value to fill censored values with, 0-9. Default 0.
```
Out of curiosity, why 0-9?

crispy-wonton replied:
because I believe values that are below 10 get censored out by ONS.

helloaidank replied:
Probably good to mention that to those who don't know

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+
+    Returns:
+        pl.Dataframe: counts of property type for all LSOAs in England and Wales
+    """
```
```suggestion
    """
    # Validate censored_values parameter
    if not (0 <= censored_values <= 9):
        raise ValueError("censored_values must be between 0 and 9 inclusive.")
```

crispy-wonton replied:
good idea :)

crispy-wonton replied:
I went with a warning rather than error in the end

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    )
+    df = df.rename({"Area Code": "lsoa"}).drop(["Area Name"])
+
+    return _fill_df_censored_values(df, censored_values)
```
Again, I would suggested putting this operation before, mostly for de-bugging reasons!

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+
+    Returns:
+        pl.Dataframe: counts of tenure type for all LSOAs in England and Wales
+    """
```
```suggestion
    """
    # Validate censored_values parameter
    if not (0 <= censored_values <= 9):
        raise ValueError("censored_values must be between 0 and 9 inclusive.")
```

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+        }
+    ).drop(["Area Name"])
+
+    return _fill_df_censored_values(df, censored_values)
```
same as before!

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    df = df.rename(df[2].to_dicts().pop()).slice(
+        3,
+    )
+    df = df.rename(
```
Add separate dictionary/list with what you're renaming here for readability.

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    return _fill_df_censored_values(df, censored_values)
+
+
+def get_df_target_tenure_uncensored() -> pl.DataFrame:
```
What do you mean by uncensored here?

crispy-wonton replied:
this function gets target tenure dataset (same as `get_df_target_tenure() `) but this time uses a different dataset which doesnt have any censored values for low numbers. the trade off is that the categories had to be manually aggregated and the code is longer. have added to docstring

I kept both functions for now in case it makes a difference in the pipeline. we can delete one later or keep both.

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    post_cols: list = config["mapping"]["build_year_post_cols"],
+) -> pl.DataFrame:
+    """
+    Get dataframe of counts of properties built before and after given year for all LSOAs in England and Wales. Source:
```
Probably need a bit more info in this doc string about the range of years that can be used. I would potentially think I would be able to use any year here but that's not the case right?

crispy-wonton replied:
have updated docstring - see what you think. The split year in this pipeline will depend on the age bands in both target and sample datasets but this specific function will work with any age bands available in this target dataset - so yes, not any year.

helloaidank commented on asf_heat_pump_suitability/getters/get_target.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    cols = df.columns
+    cols.remove("lsoa")
+    return df.with_columns([pl.col(cols).str.replace("c", f"{val}").cast(pl.Int64)])
```
Same as before :)

helloaidank commented:
Hi @crispy-wonton , I've reviewed your code! Looks great, all very clean and understandable. Just a couple of comments on potentially chaining polar commands together, not performing operations in the return part of the function (mostly for readability and debugging) and putting some of the string heavy parameters into lists/dictionaries. 

Well done 🚀 !

crispy-wonton committed:
make changes requested in review
- clean return statements of functions
- raise warning if fill_censored value is outside suggested range
- update docstrings of getters/get_target.py functions for clarity

crispy-wonton committed:
add explanation of censored values in getter/get_target.py functions

crispy-wonton committed:
set weight and id columns in pipeline/reweight_epc/prepare_target.generate_balance_target_population

crispy-wonton committed:
Merge pull request #22 from nestauk/20_prepare_target_reweighting

20 prepare target reweighting

add new pipeline/reweight_epc/prepare_target.py file with functions to prepare target marginals for reweighting with balance

crispy-wonton committed:
Merge branch 'dev' into 13_add_target_data_getters
